### ISSUE TITLE
Automatic Named Entity Recognition

### ISSUE THREAD

**Issue by ChiaraPalladino** [pelagios/recogito2#126]
There I am back with new issues...started testing again for your joy. 
I don't know if it's supposed to be active now, but automatic NER does not work. I tested it on the Itinerarium Antonini, and it usually could recognize up to 20 places (considering the size of the itinerary, its a risible quantity). I now uploaded the Greek document I geoannotated entirely on v. 1, but it didn't recognize a thing. 
Should I test it on English as well? And most importantly - is it supposed to work now?

**Comment by rsimon** [comment]
We're using Stanford NLP underneath the hood. English is the only language that's going to work properly for the time being. Since we can't fix anything about the NER per se (we need to rely on what other libs can do), it "works as designed" as far as Recogito is concerned I'm afraid.

If something completely breaks with an error, or NER isn't even started, that that would be another thing (under our control). But that didn't seem to happen in your case. (Since it did recognize at least a few places). Things are hopefully different once we can try out integrating the CLTK #92. But I can't give a fixed date when that will happen :-(

**Comment by ChiaraPalladino** [comment]
Just tested on the English translation. Now _this_ is weird...

![image](https://cloud.githubusercontent.com/assets/15249889/16523525/397af94a-3fa4-11e6-8c5a-a9a7725d44ff.png)

It seems not to recognize capitalized letters....

**Comment by rsimon** [comment]
Argh - hm. I don't think it's a problem with the NER. Rather, I think, something's causing a one-character offset in the annotations. Can you share the doc with me so I can take a look? (Go to the settings menu icon, then to "Sharing", and under "collaborators" search for "rainer" and give me read permissions.

**Comment by rsimon** [comment]
Ah damn - seems that new sharing feature is currently broken. I'll get back to you about that!

**Comment by ChiaraPalladino** [comment]
Just noticed! - it used to work though, I saw you shared Odyssey and Bourdeaux Itinerary.

**Comment by rsimon** [comment]
Ok, that's pretty weird. It seems it does work, but an older (previously broken) version might still have been cached in our browsers. Can you try to hit SHIFT+Refresh?

**Comment by ChiaraPalladino** [comment]
There I am! 
I got slightly confused at the document sharing options, because actually you should be able to add collaborators either by clicking on "document settings and metadata" > "sharing" and by "invite other users" (which I imagine is not yet active). I included you as an administrator for the English document, but I don't know if I succeeded.

**Comment by ChiaraPalladino** [comment]
Anyway, NER and georeference in English works pretty well besides this "bug". It got 182 place names and also georeferenced the majority of them (also quite better than v. 1 as far as I can see).

**Comment by rsimon** [comment]
Sorry! The "invite other users" main menu option will actually be removed; Settings->Sharing is the only place you'll be able to share - I should have told you! (Or rather I should have removed the icon altogether. In it's place, there will be a "discussion board" icon. I'll make that change ASAP.)

Anyways, I can access the doc, so that seems to work fine. I can't spot the reason for the offset right away. But rest assured I'll track that bug down ;-)

**Comment by rsimon** [comment]
Speaking of off-by-one errors: https://twitter.com/codinghorror/status/506010907021828096

**Comment by rsimon** [comment]
I'm getting a rough idea about what's going on. In your particular case, it had to do with leading empty lines. These, obviously, are not supposed to break things like. I'll need to set up a few tests next weeks to find out where exactly things start running out of sync. But now that I know what I'm looking for, I'll be able to sort this out. Thx for spotting!

**Comment by ChiaraPalladino** [comment]
Will keep an eye on this and see if there are similar bugs around...

**Comment by rsimon** [comment]
Hi Chiara,

I deployed a fix for this (which, in fact, turned out to be quite minimal). Newlines shouldn't cause any problems any more. Special characters, however, could in some cases. I'm working on this next.

Cheers,
R

**Comment by ChiaraPalladino** [comment]
I was very curious so had a look at it immediately...WOW! 

I can't recall many instances of special characters in English, at least, but working with OCR (especially bad OCR) could cause this problem perhaps.

**Comment by rsimon** [comment]
The issue remains a bit puzzling... it was caused by the use of a special HTML tag that's supposed to display pre-formatted text, the way it's structured (newlines, paragraphs) in the original source. Turned out that it was this HTML tag which didn't properly deal with the newlines, and added the offset.

Once I replaced it with a generic DIV structure tag, things worked just fine. And an additional style rule - which I already had in there (I frankly forgot why...) is still keeping the original formatting intact. 

Concerning special characters (issue #122): checked this now; and everything should be clean there, too. At least if the source text is proper UTF-8, special characters should display fine; and every character that requires special treatment by the rules of HTML (&, angle brackets etc.) is properly handled by the page rendering process as it's implemented. So I'm keeping my finger's crossed that the UI is now operational ;-)
